### ISSUE TITLE
Bad access error when deserializing a fully connected TF Lite model

### ISSUE THREAD

**Issue by ushnish** [tensorflow/tensorflow#15488]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **TensorFlow installed from (source or binary)**:
- **TensorFlow version (use command below)**:
- **Python version**: 
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"
`('v1.3.0-rc1-5910-ge2174cc943', '1.4.0')`

### Describe the problem
I trained a very simple feed forward network (model structure is y = W*x + b so basically linear regression) using raw tensorflow code (matrix multiply, bias add and relu) so not using any Estimators or higher order APIs. My main purpose was to see if I could serialize the model, convert it to .lite format and then deserialize it correctly in C++.

I did freeze_graph on this graph.pbtxt and model checkpoints using tensorflow's provided freeze_graph method, and converted to .lite format using the bazel command line instruction (in the lite documentation) Lite conversion command used is

```
bazel run --config=opt \
  //tensorflow/contrib/lite/toco:toco -- \
  --input_file=frozen_model.pb  \               
 --output_file=frozen_model.lite \               
 --input_format=TENSORFLOW_GRAPHDEF \
  --output_format=TFLITE \
  --inference_type=FLOAT \
  --input_shape=**1,1** \ 
 --input_array=**x** \          
 --output_array=**output**
``` 
As x is the name of the input 1-Dim placeholder tensor, output is the name of the result of W*x+b. I did not include y as part of the input_array as running with --input_array=x,y, --input_shape=1,1:1,1 as this produced a .lite model with 0 bytes (no output)

Using code very similar to what is in tflite_driver.cc and tflite_driver_test.cc, I was able to get the mobile_net example working fine but this simple model I described above fails with a EXC_BAD_ACCESS error. Upon debugging I saw that the error happens at fully_conncted.cc

```
template <KernelType kernel_type>
TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
....
  switch (input->type) {  // Already know in/out types are same.
    case kTfLiteFloat32:
      return EvalFloat<**kernel_type**>(context, node, params, data, input, filter,
                                    bias, output);
....
  }
  return kTfLiteOk;
}

```

kernel_type is **kPie** at this point. Looking further into fully_connected.cc, I notice two things.

```
enum KernelType {
  kReference,
  kGenericOptimized,  // Neon-free
  kNeonOptimized,
  kPie,  **// Used by the PIE team**
};

```
And also

```
TfLiteRegistration* Register_FULLY_CONNECTED() {
  **// TODO(ahentz): We don't have a dedicated quantized version of the PIE
  // kernel. For now, the quantized version just defer to the corresponding
  // optimized MINI kernel. At some point we will allow different libraries to
  // be built with different kernels, but for now we have to pick one here.
  return Register_FULLY_CONNECTED_PIE();**
#ifdef USE_NEON
  return Register_FULLY_CONNECTED_NEON_OPT();
#else
  return Register_FULLY_CONNECTED_GENERIC_OPT();
#endif
}

```
Specifically, the EXC_BAD_ACCESS error happens at

```
TfLiteStatus EvalPie(TfLiteContext* context, TfLiteNode* node,
                     ...
  // Output = bias if bias tensor exists.
  if (bias) {
    tensor_utils::VectorBatchVectorAssign(bias->data.f, num_units, batch_size,
                                          output->data.f);
  } 
```

![upload](https://user-images.githubusercontent.com/3603839/34171209-63703e96-e4a2-11e7-89fd-446b1c1c44cb.png)

As we can see the filter and bias terms have bad memory addresses, the input and output tensors have valid memory addresses though

My questions are 
1) Why is kPie kernel_type being picked for my simple model?
2) Based on the comments it looks like kPie is not supported for external use, so could this explain the bad access error?
3) Is this an issue with how I converted to lite format? Here is my command again
```
bazel run --config=opt \
  //tensorflow/contrib/lite/toco:toco -- \
  --input_file=frozen_model.pb  \               
 --output_file=frozen_model.lite \               
 --input_format=TENSORFLOW_GRAPHDEF \
  --output_format=TFLITE \
  --inference_type=FLOAT \
  --input_shape=**1,1** \ 
 --input_array=**x** \          
 --output_array=**output**
``` 
4) Is there any example of a very simple model composed out of matmul, add and relu which can be converted to .lite in such a way that it can be correctly deserialized in C++, just like the mobile_net example? I can build up from such an example

**Comment by ushnish** [comment]
Hey guys, wanted to add an update to this as I tried to get something much simpler to work. Please let me know if there is something not quite right in what I attempted.

I installed TF 1.5 from source.

This is the simplest possible DNN model - no hidden layer, dropout, or activation function, bias and weight are 1 dimensional.

`import tensorflow as tf
import numpy as np
from tensorflow.python.framework.graph_util import convert_variables_to_constants

model_dir = "simple_lr_ckpt/"
x = tf.placeholder(tf.float32, shape=[None,1], name="x")
y = tf.placeholder(tf.float32, shape=[None,1], name="y")
W = tf.Variable(tf.truncated_normal(shape=[1, 1], mean=0, stddev=0.001, dtype=tf.float32), name="W")
b = tf.Variable(tf.truncated_normal(shape=[1], mean=0, stddev=0.001, dtype=tf.float32), name="b")
output_name = "logit"
logit = tf.add(tf.matmul(x, W), b, name=output_name)
loss = tf.reduce_mean(tf.square(logit - y))
tf.summary.scalar('loss', loss)
train_op = tf.train.AdamOptimizer(learning_rate=0.001).minimize(loss)
merged = tf.summary.merge_all()
N = 1000000
batch_size = 100
saver = tf.train.Saver()
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    train_writer = tf.summary.FileWriter(model_dir, sess.graph)
    for start in xrange(0, N, batch_size):
        x_ = np.reshape(np.random.rand(batch_size), (-1,1))
        y_ = 2*x_ + 1
        _, mse, summary = sess.run([train_op, loss, merged], feed_dict={x: x_, y: y_})
        if start % (batch_size*100) == 0:
            print mse
        # if start % batch_size * 100 == 0:
            saver.save(sess, model_dir + "model.ckpt-old", global_step=start)
        train_writer.add_summary(summary, start)

    tf.train.write_graph(sess.graph_def, model_dir, 'graph.pbtxt')
    frozen_graph_def = convert_variables_to_constants(sess, sess.graph_def, [output_name])

    with tf.gfile.GFile(model_dir+ "graph.pb", "w") as f:
        f.write(frozen_graph_def.SerializeToString())`

To freeze the graph this is the code I used

`import os
from tensorflow.python.tools import freeze_graph
import tensorflow as tf

dir(tf.contrib)
output_path = model_dir
input_graph_path = os.path.join(output_path, "graph.pbtxt")
input_saver_def_path = ""
input_binary = False
restore_op_name = "save/restore_all"
filename_tensor_name = None
output_graph_path = os.path.join(output_path, "frozen_graph.pb")
clear_devices = False
freeze_graph.freeze_graph(
    input_graph_path, input_saver_def_path, input_binary, tf.train.latest_checkpoint(model_dir),
    output_name, restore_op_name, filename_tensor_name,
    output_graph_path, clear_devices, "", "")`

After this, I convert the .pb file to .lite using the following

`bazel-bin/tensorflow/contrib/lite/toco/toco \
  --input_format=TENSORFLOW_GRAPHDEF \                                                                        
  --input_file=/Users/ushnishde/tf_lite_example_new/simple_lr_ckpt/frozen_graph.pb \
  --output_format=TFLITE \
  --output_file=/Users/ushnishde/tf_lite_example_new/simple_lr_ckpt/frozen_graph.lite \
  --inference_type=FLOAT \
  --inference_input_type=FLOAT \
  --input_arrays=x \
  --output_arrays=logit \
  --input_shapes=1,1
`

I kept the names of the tensors consistent from when I created the model. Is the y tensor considered as one of the input_arrays?


When I try to deserialize the .lite model in C++, I followed the code statements which ultimately lead to the EXC_BAD_ACCESS error

tflite::Interpreter::Invoke() interpreter.cc:414
<img width="522" alt="screen shot 2018-01-03 at 12 32 01 am" src="https://user-images.githubusercontent.com/3603839/34513536-8fc5d04c-f01d-11e7-9ffa-df64ccf53545.png">

TfLiteStatus tflite::ops::builtin::fully_connected::Eval<(tflite::ops::builtin::fully_connected::KernelType)3>(TfLiteContext*, TfLiteNode*) fully_connected.cc:250
<img width="664" alt="screen shot 2018-01-03 at 12 32 37 am" src="https://user-images.githubusercontent.com/3603839/34513554-a37ad13c-f01d-11e7-8a05-b3768ac7612e.png">

TfLiteStatus tflite::ops::builtin::fully_connected::EvalFloat<(tflite::ops::builtin::fully_connected::KernelType)3>(TfLiteContext*, TfLiteNode*, TfLiteFullyConnectedParams*, tflite::ops::builtin::fully_connected::OpData*, TfLiteTensor*, TfLiteTensor*, TfLiteTensor*, TfLiteTensor*) fully_connected.cc:226
<img width="652" alt="screen shot 2018-01-03 at 12 33 15 am" src="https://user-images.githubusercontent.com/3603839/34513573-b61ae930-f01d-11e7-9fa5-c58c6ba74d14.png">

tflite::ops::builtin::fully_connected::EvalPie(TfLiteContext*, TfLiteNode*, TfLiteFullyConnectedParams*, tflite::ops::builtin::fully_connected::OpData*, TfLiteTensor*, TfLiteTensor*, TfLiteTensor*, TfLiteTensor*) fully_connected.cc:148
<img width="647" alt="screen shot 2018-01-03 at 12 33 45 am" src="https://user-images.githubusercontent.com/3603839/34513588-c9a6ea08-f01d-11e7-8b07-16b40857b3fe.png">

tflite::tensor_utils::VectorBatchVectorAssign(float const*, int, int, float*) portable_tensor_utils.h:147
<img width="513" alt="screen shot 2018-01-03 at 12 34 16 am" src="https://user-images.githubusercontent.com/3603839/34513599-dca703d6-f01d-11e7-8bed-b5a83ab9fa74.png">

tflite::tensor_utils::PortableVectorBatchVectorAssign(float const*, int, int, float*) portable_tensor_utils.cc:104
<img width="524" alt="screen shot 2018-01-03 at 12 34 54 am" src="https://user-images.githubusercontent.com/3603839/34513615-f2a41b24-f01d-11e7-87c4-ef2be1638d84.png">

`void PortableVectorBatchVectorAssign(const float* vector, int v_size,
                                     int n_batch, float* batch_vector) {
  for (int b = 0; b < n_batch; b++) {
    memcpy(batch_vector + b * v_size, vector, v_size * sizeof(float));
  }
}`
<img width="554" alt="screen shot 2018-01-03 at 12 35 48 am" src="https://user-images.githubusercontent.com/3603839/34513633-10fa2c3a-f01e-11e7-93de-7bcdf1a23f87.png">

Any ideas what could be going on? I'm eager to try any suggestions at this point.

**Comment by ushnish** [comment]
@aselle any ideas on how we can proceed forward? Thank you.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by andrehentz** [comment]
@ushnish We have been unable to reproduce this. It seems to work on Android1, Nexus5 and Nexus-6p phones. Is there anything special about your setup?

**Comment by ushnish** [comment]
Hello Andre, thanks for responding. I am working on a MacBook Pro, High Sierra. I use CLion as my IDE.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by aselle** [comment]
@ushnish, were you able to resolve your issue?

**Comment by ushnish** [comment]
Unfortunately I was not able to resolve this issue. Has anyone tried to run the end-to-end example I provided? I’ll be happy to revisit this issue if the example I provided works now

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 77 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 92 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by andrehentz** [comment]
I just went over your example, and it should work without issues. 

My guess is that your inference code is not holding on to the flatbuffer bytes (or of the FlatBufferModel) for the whole duration of inference. 

Here's what I've done:

- Conversion:
```
bazel run tensorflow/contrib/lite/toco:toco -- --input_format=TENSORFLOW_GRAPHDEF \
   --input_file=frozen_graph.pb --inference_type=FLOAT --inference_input_type=FLOAT \
   --input_arrays=x --output_arrays=logit --input_shapes=1,1 \
   --output_file=frozen_graph.tflite
```
- Accuracy (vs TF):
```
bazel run tensorflow/contrib/lite/testing:tflite_diff --  --tensorflow_model=frozen_graph.pb \
   --tflite_model=frozen_graph.tflite --input_layer=x --input_layer_type=float \
   --input_layer_shape=1,1 --output_layer=logit
```
- Latency
```
bazel run tensorflow/contrib/lite/tools/benchmark:benchmark_model -- \
    --num_threads=4 --num_runs=1000  --graph=frozen_graph.tflite
```
